Add pyright as well as tests and infrastructure inspired by bugbear

jakkdl opened the pull request:
I didn't find a way to configure pyright in `setup.py`, opted to use a `pyproject.toml` instead of a `pyrightconfig.json`.

With `pytest-xdist` commented out passing `-n auto` errors out. Commented out the parameter, but probs cleaner to uncomment both lines or remove both.

Could opt to explicitly list files to avoid checking `setup.py` instead of adding the check for `__version__` in it.

jakkdl commented:
All checks are for stuff I didn't touch in this pull request, opening another one that works on `flake8_trio.py` and adds basic tests.

Zac-HD commented:
- pyright config lgtm, discovery is less error-prone than explicit lists
- let's leave xdist commented out for now; we'll uncomment once we have ~10 tests and concurrency is worth the overhead
- obvious test failures are obvious, feel free to fix them here or in another PR as you prefer.

jakkdl commented:
kk, let's just do that here then.
I haven't delved into understanding the testing setup that bugbear uses, so just doing basic tests for now.

Zac-HD commented on flake8_trio.py:
```diff
@@ -34,15 +33,19 @@ def is_trio_call(node: ast.AST, *names: str) -> Optional[str]:
+            for item in node.items
+        ]
+        if any(calls) and not any(isinstance(x, ast.Await) for x in ast.walk(node)):
+            self.problems.append(
+                (node.lineno, node.col_offset, TRIO100.format(calls.pop()))
+            )
```
Actually important to use the `item` attributes as in my previous loop, so that we point at the call to `trio.xxx_after()` if there are multiple clauses e.g.
```python
with (
    trio.move_on_after(10),
    httpx.AsyncContext() as client,
): ...
```

jakkdl replied:
`ast.withitem` doesn't have a `lineno` - so it's afaict not possible, changed it because it threw errors at me in tests.

Yeah the Bugbear way sounds good, will look into the boilerplate needed.

jakkdl replied:
ah, the `context_expr` in the `withitem` has a lineno though.

Zac-HD reviewed (commented):
LGTM with a comment below - I would recommend flake8-bugbear testing, it amounts to "write a Python file, then get results for that file, and assert that they're the expected results".  Just a bit easier to read than embedding source code as string literals!

jakkdl commented:
Okay, cribbed bugbear's testing setup. Had to replace the `namedtuple`+`partial` solution to get it to pass pyright, though likely possible to get a cleaner solution.
Dropped stdin support for the `filename` parameter that bugbear had, was way too much of a hassle to get that covered and I'm not sure how big the need for it is.
I've never done fuzz tests before, but copy-pasted to my heart's content and they seem to do something.
Added back xdist since fuzzers started taking time.

Zac-HD commented on tests/test_flake8_trio.py:
```diff
@@ -0,0 +1,66 @@
+        self.assertEqual(errors, expected)
+
+
+class TestFuzz(unittest.TestCase):
```
I'd skip this by default - as you note it's pretty slow, and I think running it by hand occasionally is plenty.

jakkdl replied:
Adding a `pytest.mark.fuzz` to them, so could maybe add `--runfuzz` to the release test (which I think only you can add?).

Zac-HD commented on flake8_trio.py:
```diff
@@ -34,29 +33,68 @@ def is_trio_call(node: ast.AST, *names: str) -> Optional[str]:
+            yield problem.values()
+
+
+class Error:
+    def __init__(self, lineno: int = -1, col: int = -1, message: str = ""):
+        self.lineno = lineno
+        self.col = col
+        self.message = message
+        self.err_type = type(Plugin)
+
+    def values(self):
+        return (self.lineno, self.col, self.message, self.err_type)
+
+
+class TRIO100(Error):
```
I don't think this is worth the overhead when we're just going to work with tuples anyway.

If the iteration I had was unsatisfying, how about we include `Plugin` in each tuple in `Visitor.problems`, so that we can just `yield from visitor.problems`?

jakkdl replied:
It was because I was hitting typing problems with the `partial` solution that bugbear had, but yeah I simplified it down a lot and am just doing a tuple + factory function.

Lots of minor variations this could take, could do a `namedtuple` (but atm never accesses any members anyway so w/e), or could do away with the function and let callers do the format + run add the type.

Zac-HD commented on flake8_trio.py:
```diff
@@ -34,29 +33,68 @@ def is_trio_call(node: ast.AST, *names: str) -> Optional[str]:
-    def run(self) -> Generator[tuple[int, int, str, Type[Any]], None, None]:
+        Stolen from flake8_import_order because it's good.
+        """
+        lines = pycodestyle.readlines(filename)
+        return ast.parse("".join(lines))
```
```suggestion
        with tokenize.open(filename) as f:
            return ast.parse(f.read())
```

I'm not actually a fan of the "on error assume UTF-8" approach; it seems likely to cause syntax problems later and doesn't actually fix the encoding error.

jakkdl replied:
This one I more or less just straight copy-pasted w/o much investigation. Sounds good, and yeah it looked quite ugly after I skipped out on `stdin` support.

Zac-HD commented on tests/test_flake8_trio.py:
```diff
@@ -0,0 +1,66 @@
+        self.assertEqual(errors, [])
+
+    def test_trio100(self):
+        filename = Path(__file__).absolute().parent / "trio100.py"
+        plugin = Plugin(filename=str(filename))
+        errors = list(plugin.run())
```
Let's wrap this up in a helper `def get_errors_for_filename(fname: str) -> ast.AST:`, since we'll be doing it a lot.

Or even `def assert_expected_errors(self, fname: str, *expected: Tuple[...]) -> None:` to cover *all* the boilerplate.

jakkdl replied:
yeah was definitely gonna do that sooner or later, but I thought I might wait until I have a few more tests implemented to see how much of the work is identical between them and the best way to lay out the helper function. But I can push in an initial version now

jakkdl commented:
fixed all comments, code starting to look pretty :sparkle:

Zac-HD reviewed (approved):
Lovely!  Merging this now since it's a huge improvement on the status quo (😅), and I'll push a small refactoring commit in lieu of nitpicky review.
